orca-webhook/src/main/groovy/com/netflix/spinnaker/orca/webhook/tasks/MonitorWebhookTask.groovy

```diff
@@ -53,48 +53,54 @@ class MonitorWebhookTask implements OverridableTimeoutRetryableTask {
   @Autowired
   WebhookService webhookService
 
-  static requiredParameters = ["statusEndpoint", "statusJsonPath"]
-
   @Override
   TaskResult execute(Stage stage) {
-    def missing = requiredParameters.findAll { !stage.context.get(it) }
-    if (!missing.empty) {
-      throw new IllegalStateException("Missing required parameter${missing.size() > 1 ? 's' : ''} '${missing.join('\', \'')}'")
-    }
+    StageData stageData = stage.mapTo(StageData)
 
-    String statusEndpoint = stage.context.statusEndpoint
-    String statusJsonPath = stage.context.statusJsonPath
-    String progressJsonPath = stage.context.progressJsonPath
-    String successStatuses = stage.context.successStatuses
-    String canceledStatuses = stage.context.canceledStatuses
-    String terminalStatuses = stage.context.terminalStatuses
-    def customHeaders = stage.context.customHeaders
+    if (stageData.statusEndpoint == null || stageData.statusJsonPath == null) {
+      throw new IllegalStateException(
+        "Missing required parameter(s): statusEndpoint = ${stageData.statusEndpoint}, statusJsonPath = ${stageData.statusJsonPath}")
+    }
 
     def response
     try {
-      response = webhookService.getStatus(statusEndpoint, customHeaders)
+      response = webhookService.getStatus(stageData.statusEndpoint, stageData.customHeaders)
       log.debug(
         "Received status code {} from status endpoint {} in execution {} in stage {}",
         response.statusCode,
-        statusEndpoint,
+        stageData.statusEndpoint,
         stage.execution.id,
         stage.id
       )
     } catch (IllegalArgumentException e) {
       if (e.cause instanceof UnknownHostException) {
-        log.warn("name resolution failure in webhook for pipeline ${stage.execution.id} to ${statusEndpoint}, will retry.", e)
+        log.warn("name resolution failure in webhook for pipeline ${stage.execution.id} to ${stageData.statusEndpoint}, will retry.", e)
         return TaskResult.ofStatus(ExecutionStatus.RUNNING)
       }
 
-      throw e
+      String errorMessage = "an exception occurred in webhook monitor to ${stageData.statusEndpoint}: ${e}"
+      log.error(errorMessage, e)
+      Map<String, ?> outputs = [webhook: [monitor: [:]]]
+      outputs.webhook.monitor << [error: errorMessage]
+      return TaskResult.builder(ExecutionStatus.TERMINAL).context(outputs).build()
     } catch (HttpStatusCodeException  e) {
       def statusCode = e.getStatusCode()
-      if (statusCode.is5xxServerError() || statusCode.value() == 429) {
-        log.warn("error getting webhook status from ${statusEndpoint}, will retry", e)
+      def statusValue = statusCode.value()
```

Comment:
I wonder if there's a way to pull this logic below when we map HTTP statuses onto task statuses.  It's a bit confusing that we do some status mapping up here in a catch block, and then other mapping below...which is pre-existing though now we're adding more logic here.

Reply:
will take a peek tomorrow

Reply:
yeah, i hear you - this method has about 10 too many `return`s... Not sure how to cleanly refactor these without thinking a bunch more about it. Do you have some ideas that come to you?

Reply:
Maybe just store the status in this block, then fall back to the status logic below?  Though I'm not sure if any of the intermediate logic still running would be a problem.  If it's not a simple change, no worries...just an idea to improve it if reasonable in scope.

Reply:
the problem is that then you can't run all the "response.xxx" code. maybe something for next time

Reply:
Sounds good!